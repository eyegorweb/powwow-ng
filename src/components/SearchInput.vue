<template>
  <div>
    <input
      type="text"
      v-model="$value"
    >
    <div>
      <slot
        :results="results"
        :query="$value"
      />
    </div>
  </div>
</template>

<script>
import { propWithDataFallback } from 'vue-prop-data-fallback';

export default {
  // accepte une prop value mais peut marcher sans
  mixins: [propWithDataFallback('value')],

  props: {
    items: {
      type: Array,
      required: true,
    },
    fields: {
      type: Array,
      required: true,
    },
  },

  computed: {
    results() {
      return this.$value
        ? this.items.filter(item => this.fields.some(field => item[field].includes(this.$value)))
        : this.items;
    },
  },
};
</script>

<style scoped>
</style>
