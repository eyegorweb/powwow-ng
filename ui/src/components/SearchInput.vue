<template>
  <div class="search-input">
    <UiInput
      v-model="$value"
      :block="block"
      :placeholder="placeholder"
      @update:value="updateValue"
      @click="suggestionsAreVisible = true"
    >
      <!-- TODO:Utiliser la bonne icone -->
      <img style="font-size: 24px" slot="icon" src="@/assets/search.svg" :style="{ left: 0 }" />
      <template slot="beforeInput">
        <slot name="beforeInput" />
      </template>
    </UiInput>

    <div
      v-if="suggestionsAreVisible"
      :class="{ collapsed: collapsedMode }"
      v-clickaway="hideIfCollapsedMode"
    >
      <slot :results="results" :query="$value" />
    </div>
  </div>
</template>

<script>
import { propWithDataFallback } from 'vue-prop-data-fallback';
import UiInput from '@/components/ui/UiInput';
import fuzzysort from 'fuzzysort';
import { clickaway } from '@/directives/clickaway';
import { containsWithHighlight } from '@/utils.js';

export default {
  // accepte une prop value mais peut marcher sans
  mixins: [propWithDataFallback('value')],

  props: {
    items: {
      type: Array,
      required: true,
    },
    fields: {
      type: Array,
      required: true,
    },
    collapsedMode: Boolean,
    placeholder: {
      type: String,
      default: '',
      required: false,
    },
    block: Boolean,
    containsSearch: Boolean,
  },
  directives: { clickaway },

  components: {
    UiInput,
  },

  computed: {
    results() {
      if (!this.$value) return this.highlightedResults;

      if (this.containsSearch) {
        return containsWithHighlight(this.$value, this.items);
      } else {
        return fuzzysort
          .go(this.$value, this.items, {
            keys: this.fields,
            allowTypo: false,
          })
          .map(r => ({
            item: r.obj,
            highlighted: this.fields.reduce((highlighted, field, i) => {
              highlighted[field] = fuzzysort.highlight(r[i]) || r.obj[field];
              return highlighted;
            }, {}),
          }));
      }
    },

    // version non surlignée à autiliser losque la recherche est vide
    highlightedResults() {
      if (!this.items) return [];

      return this.items.map(item => ({
        item,
        highlighted: this.fields.reduce((highlighted, field) => {
          highlighted[field] = item[field];
          return highlighted;
        }, {}),
      }));
    },
  },

  data() {
    return {
      suggestionsAreVisible: true,
    };
  },

  mounted() {
    this.suggestionsAreVisible = !this.collapsedMode;
  },

  methods: {
    hideIfCollapsedMode() {
      if (this.collapsedMode) {
        this.suggestionsAreVisible = false;
      }
    },
    updateValue(value) {
      if (this.collapsedMode) {
        this.suggestionsAreVisible = value && value.length;
      }
      this.$emit('update:value', value);
    },
  },
};
</script>

<style lang="scss" scoped>
.search-input {
  position: relative;

  .close-icon {
    position: absolute;
    width: 1em;
    top: 0.5rem;
    right: 0;

    &:hover {
      cursor: pointer;
    }
  }
}

.collapsed {
  position: absolute;
  width: 100%;
  background: white;
  z-index: 999;
  border: 1px solid $gray-400;
  border-radius: 3px;
  padding-left: 10px;
  border-top: none;
}
</style>
