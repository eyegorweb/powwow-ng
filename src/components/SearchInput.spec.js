import { shallowMount } from '@vue/test-utils';
import SearchInput from './SearchInput.vue';

describe('SearchInput.vue', () => {
  /** @type {import('@vue/test-utils').Wrapper} */
  let wrapper;
  beforeEach(() => {
    wrapper = shallowMount(SearchInput, {
      propsData: {
        value: '',
        items: [{ name: 'banana' }, { name: 'apple' }, { name: 'orange' }],
        fields: ['name'],
      },
      scopedSlots: {
        default: `<div class="results" slot-scope="{ results }">
          <span class="result" v-for="result in results">{{ result.name }} </span>
        </div>`,
      },
    });
  });

  it('gives all results with empty value', () => {
    expect(wrapper.find('.results').text()).toBe('banana apple orange');
  });

  it('filters results', () => {
    wrapper.setProps({ value: 'nana' });
    expect(wrapper.find('.results').text()).toBe('banana');
  });
});
