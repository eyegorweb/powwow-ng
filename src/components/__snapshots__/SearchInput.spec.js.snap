// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchInput.vue multi field displays highlighted results 1`] = `
<div class="results-hl"><span class="result-hl">{
  "firstName": "Jonathan",
  "lastName": "Joe&lt;b&gt;star&lt;/b&gt;"
}</span><span class="result-hl">{
  "firstName": "Joseph",
  "lastName": "Joe&lt;b&gt;star&lt;/b&gt;"
}</span><span class="result-hl">{
  "firstName": "Jotaro",
  "lastName": "Joe&lt;b&gt;star&lt;/b&gt;"
}</span></div>
`;

exports[`SearchInput.vue multi field displays highlighted results when no value provided 1`] = `
<div class="results-hl"><span class="result-hl">{
  "firstName": "Jonathan",
  "lastName": "Joestar"
}</span><span class="result-hl">{
  "firstName": "Joseph",
  "lastName": "Joestar"
}</span><span class="result-hl">{
  "firstName": "Jotaro",
  "lastName": "Joestar"
}</span></div>
`;

exports[`SearchInput.vue multi field gives all results with empty value 1`] = `<div class="results"><span class="result">Jonathan Joestar - </span><span class="result">Joseph Joestar - </span><span class="result">Jotaro Joestar - </span></div>`;

exports[`SearchInput.vue single field displays highlighted results 1`] = `
<span class="result-hl">{
  "name": "ba&lt;b&gt;nana&lt;/b&gt;"
}</span>
`;

exports[`SearchInput.vue single field displays highlighted results when no value provided 1`] = `
<div class="results-hl"><span class="result-hl">{
  "name": "apple"
}</span><span class="result-hl">{
  "name": "banana"
}</span><span class="result-hl">{
  "name": "orange"
}</span></div>
`;
